### PR TITLE
Adding support for DSCP configuration

### PR DESCRIPTION
**Pull request by karthiksundaravel**
Support dscp configuration using os-net-config templates. It uses pyroute2 modules to send/receive, encode/decode the netlink messages. The persistence of the configuration is done with dcb_config.service

### CODE REVIEW

**Review by rjarry — CHANGES_REQUESTED** [review]

File: os_net_config/utils.py [hunk 1]
```diff
@@ -502,6 +534,23 @@ def nicpart_udev_rules_check():
...
+
+def configure_dcb_config_service():
+    """Generate the sriov_config.service
+
+     sriov_config service shall configure the numvfs for the SriovPF nics
+     during reboot of the nodes.
+    """
+    with open(_DCB_CONFIG_SERVICE_FILE, 'w') as f:
+        f.write(_DCB_CONFIG_DEVICE_CONTENT)
+    processutils.execute('systemctl', 'enable', 'dcb_config')
```
[the remark this answers is not in the document]
Comment (karthiksundaravel) [reply os_net_config/utils.py:551]: These services are expected to run during reboots, so IMHO its not necessary, isn't it ?

File: os_net_config/utils.py [hunk 2]
```diff
@@ -43,6 +43,22 @@
 WantedBy=basic.target
 """
 
+# dcb_config service shall be created and enabled so that the various
+# dcb configurations shall be done during reboot as well using
+# dcb_config.py installed in path /usr/bin/os-net-config-sriov
+_DCB_CONFIG_SERVICE_FILE = "/etc/systemd/system/dcb_config.service"
+_DCB_CONFIG_DEVICE_CONTENT = """[Unit]
+Description=DCB configuration
+After=NetworkManager
```
[the remark this answers is not in the document]
Comment (karthiksundaravel) [reply os_net_config/utils.py:52]: Agree

File: os_net_config/dcb_netlink.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        class ieee_pfc(nla):
+            pack = 'struct'
+            fields = (
+                ('pfc_cap', 'B'),
+                ('pfc_en', 'B'),
+                ('mbc', 'B'),
+                ('delay', 'H'),
+                ('requests', 'QQQQQQQQ'),
+                ('indications', 'QQQQQQQQ'),
+            )
```
[the remark this answers is not in the document]
Comment (karthiksundaravel) [reply os_net_config/dcb_netlink.py:102]: We use only a subset of the DCB attributes. Do you think it will still be fine with upstream pyroute2 ?

**Review by karthiksundaravel — COMMENTED** [review]
Thanks Robin for the review

File: os_net_config/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rjarry) [reply os_net_config/utils.py]: Ok, no need.

File: os_net_config/dcb_netlink.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rjarry) [reply os_net_config/dcb_netlink.py]: Yes, I think it should be possible. Even a partial implementation is better than nothing.

File: os_net_config/utils.py [hunk 4]
```diff
@@ -43,6 +43,23 @@
 WantedBy=basic.target
 """
 
+# dcb_config service shall be created and enabled so that the various
+# dcb configurations shall be done during reboot as well using
+# dcb_config.py installed in path /usr/bin/os-net-config-dcb
+_DCB_CONFIG_SERVICE_FILE = '/etc/systemd/system/'\
+    'os-net-config-dcb-config.service'
```
Comment (rjarry) [inline os_net_config/utils.py:50]: Are you forced to break this line by a linter?

If not, please don't, it looks ugly :)

**Review by rjarry — CHANGES_REQUESTED** [review]
Hi @karthiksundaravel 

it looks good (just a minor nit pick about string split).

Did you consider upstreaming the changes into pyroute2?

Could you squash all commits into one and write a good commit message explaining the motivation for this patch? What is it for? Examples of use, etc.

Thanks!

File: os_net_config/common.py [hunk 5]
```diff
@@ -146,6 +158,80 @@ def get_file_data(filename):
...
 
 
+def ensure_directory_presence(filepath):
+    dir_path = os.path.dirname(filepath)
+    if not os.path.exists(dir_path):
+        os.makedirs(dir_path)
+
+
+def write_yaml_config(filepath, data):
+    ensure_directory_presence(filepath)
```
Comment (rjarry) [inline os_net_config/common.py:168]: Just for reference, [`os.makedirs`](https://docs.python.org/3/library/os.html#os.makedirs) has an `exist_ok=` flag, this function is not required. You can replace it with:

```python
os.makedirs(os.path.dirname(filepath), exist_ok=True)
```

**Review by rjarry — CHANGES_REQUESTED** [review]

**Review by vcandapp — COMMENTED** [review]
LGTM

**Comment by vcandapp** [comment]
Merging this patch for DEV_COMPLETE milestone now. Will conitue to accomodate any further review comments for this PR as follow PRs

**Pull request merged**